web/packages/teleport/src/components/hooks/useInfiniteScroll.ts

```diff
@@ -14,115 +14,41 @@
  * limitations under the License.
  */
 
-import { useState } from 'react';
-import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+import { useEffect, useRef } from 'react';
 
-import {
-  ResourcesResponse,
-  UnifiedResource,
-  ResourceFilter,
-} from 'teleport/services/agents';
-import { UrlResourcesParams } from 'teleport/config';
-
-/**
- * Supports fetching more data from the server when more data is available. Pass
- * a `fetchFunc` that retrieves a single batch of data. After the initial
- * request, the server is expected to return a `startKey` field that denotes the
- * next `startKey` to use for the next request.
+/** Calls `callback` whenever the `trigger` element intersects the viewport.
+ *
+ * This hook is intended to be used in tandem with the `useKeyBasedPagination`
+ * hook. Example usage:
+ *
+ * ```ts
+ * const scrollDetector = useRef(null);
+ * const { fetch, resources } = useKeyBasedPagination({
+ *   fetchFunc: api.fetchItems,
+ *   filter,
+ * });
+ * useInfiniteScroll(scrollDetector.current, fetch);
+ *
+ * return (<>
+ *   {items.map(renderItem)}
+ *   <div ref={scrollDetector} />
+ * </>);
+ * ```
  */
-export function useInfiniteScroll<T extends UnifiedResource>({
-  fetchFunc,
-  clusterId,
-  params,
-  initialFetchSize = 30,
-  fetchMoreSize = 20,
-}: Props<T>): State<T> {
-  const { attempt, setAttempt } = useAttempt('processing');
-
-  const [fetchedData, setFetchedData] = useState<ResourcesResponse<T>>({
-    agents: [],
-    startKey: '',
-    totalCount: 0,
-  });
+export function useInfiniteScroll(trigger: Element, callback: () => void) {
+  const observer = useRef<IntersectionObserver | null>(null);
 
-  const fetchInitial = async () => {
-    setAttempt({ status: 'processing' });
-    try {
-      const res = await fetchFunc(clusterId, {
-        ...params,
-        limit: initialFetchSize,
-        startKey: '',
-      });
-
-      setFetchedData({
-        ...fetchedData,
-        agents: res.agents,
-        startKey: res.startKey,
-        totalCount: res.totalCount,
-      });
-      setAttempt({ status: 'success' });
-    } catch (err) {
-      setAttempt({ status: 'failed', statusText: err.message });
-      setFetchedData({
-        agents: [],
-        startKey: '',
-        totalCount: 0,
-      });
-    }
-  };
-
-  const fetchMore = async () => {
-    // TODO(bl-nero): Disallowing further requests on failed status is a
-    // temporary fix to prevent multiple requests from being sent. Currently,
-    // they wouldn't go through anyway, but at least we don't thrash the UI
-    // constantly.
-    if (
-      attempt.status === 'processing' ||
-      attempt.status === 'failed' ||
-      !fetchedData.startKey
-    ) {
-      return;
+  useEffect(() => {
+    if (observer.current) {
+      observer.current.disconnect();
```

Comment:
Shouldn't this be done in the cleanup function returned from this `useEffect` instead?

Reply:
You're right. (The code changed a lot, but I also took your comment into consideration.)